virtcontainers/clh.go

```diff
@@ -0,0 +1,1294 @@
+		return types.HybridVSock{}, err
+	}
+	clh.Logger().WithField("function", "generateSocket").Infof("Using hybrid vsock %s:%d", udsPath, vSockPort)
+	clh.socketPath = udsPath
+	return types.HybridVSock{
+		UdsPath: udsPath,
+		Port:    uint32(vSockPort),
+	}, nil
+}
+
+func (clh *cloudHypervisor) setupVirtiofsd(timeout int) (remain int, err error) {
+
+	sockPath, perr := clh.virtioFsSocketPath(clh.id)
+	if perr != nil {
+		return 0, perr
+	}
+
+	theArgs, err := clh.virtiofsdArgs(sockPath)
+	if err != nil {
+		return 0, err
+	}
+
+	clh.Logger().WithField("path", clh.config.VirtioFSDaemon).Info()
+	clh.Logger().WithField("args", strings.Join(theArgs, " ")).Info()
+
+	cmd := exec.Command(clh.config.VirtioFSDaemon, theArgs...)
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return 0, err
+	}
+
+	if err = cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		if err != nil {
+			clh.state.VirtiofsdPID = 0
+			cmd.Process.Kill()
+		} else {
+			clh.state.VirtiofsdPID = cmd.Process.Pid
+
+		}
+		clh.storeState()
+	}()
+
+	// Wait for socket to become available
+	sockReady := make(chan error, 1)
+	timeStart := time.Now()
+	go func() {
+		scanner := bufio.NewScanner(stderr)
+		var sent bool
+		for scanner.Scan() {
+			if clh.config.Debug {
+				clh.Logger().WithField("source", "virtiofsd").Debug(scanner.Text())
+			}
+			if !sent && strings.Contains(scanner.Text(), "Waiting for vhost-user socket connection...") {
+				sockReady <- nil
+				sent = true
+			}
+		}
+		if !sent {
+			if err := scanner.Err(); err != nil {
+				sockReady <- err
+			} else {
+				sockReady <- fmt.Errorf("virtiofsd did not announce socket connection")
+			}
+		}
+		clh.Logger().Info("virtiofsd quits")
+		// Wait to release resources of virtiofsd process
+		cmd.Process.Wait()
+
+	}()
+
+	return clh.waitVirtiofsd(timeStart, timeout, sockReady,
+		fmt.Sprintf("virtiofsd (pid=%d) socket %s", cmd.Process.Pid, sockPath))
+}
+
+func (clh *cloudHypervisor) waitVirtiofsd(start time.Time, timeout int, ready chan error, errMsg string) (int, error) {
+	var err error
+
+	timeoutDuration := time.Duration(timeout) * time.Second
+	select {
+	case err = <-ready:
+	case <-time.After(timeoutDuration):
+		err = fmt.Errorf("timed out waiting for %s", errMsg)
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	// Now reduce timeout by the elapsed time
+	elapsed := time.Since(start)
+	if elapsed < timeoutDuration {
+		timeout = timeout - int(elapsed.Seconds())
+	} else {
+		timeout = 0
+	}
+	return timeout, nil
+}
+
+func (clh *cloudHypervisor) virtiofsdArgs(sockPath string) ([]string, error) {
+
+	sourcePath := filepath.Join(kataHostSharedDir(), clh.id)
+	if _, err := os.Stat(sourcePath); os.IsNotExist(err) {
+		os.MkdirAll(sourcePath, os.ModePerm)
+	}
+
+	args := []string{
+		"-f",
+		"-o", "vhost_user_socket=" + sockPath,
+		"-o", "source=" + sourcePath,
+		"-o", "cache=" + clh.config.VirtioFSCache}
+
+	if len(clh.config.VirtioFSExtraArgs) != 0 {
```

Comment:
Should we check that VirtioFSExtraArgs contains none of the 3 args from line 665 to 667 ?